File: sdks/csharp/src/NovaBilling/Types/ChargeResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record ChargeResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("planId")]
    public required string PlanId { get; set; }

    [JsonPropertyName("billableMetricId")]
    public required string BillableMetricId { get; set; }

    [JsonPropertyName("chargeModel")]
    public required ChargeResponseChargeModel ChargeModel { get; set; }

    [JsonPropertyName("billingTiming")]
    public required ChargeResponseBillingTiming BillingTiming { get; set; }

    [JsonPropertyName("invoiceDisplayName")]
    public string? InvoiceDisplayName { get; set; }

    [JsonPropertyName("minAmountCents")]
    public double? MinAmountCents { get; set; }

    [JsonPropertyName("prorated")]
    public required bool Prorated { get; set; }

    /// <summary>
    /// Model-specific config
    /// </summary>
    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }

    [JsonPropertyName("graduatedRanges")]
    public IEnumerable<ChargeGraduatedRangeResponse> GraduatedRanges { get; set; } =
        new List<ChargeGraduatedRangeResponse>();

    [JsonPropertyName("filters")]
    public IEnumerable<ChargeFilterResponse> Filters { get; set; } =
        new List<ChargeFilterResponse>();

    [JsonPropertyName("createdAt")]
    public required string CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public required string UpdatedAt { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/ruby/lib/novabilling/types/charge_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class ChargeResponse < Internal::Types::Model
      field :id, -> { String }, optional: false, nullable: false
      field :plan_id, -> { String }, optional: false, nullable: false, api_name: "planId"
      field :billable_metric_id, -> { String }, optional: false, nullable: false, api_name: "billableMetricId"
      field :charge_model, -> { Novabilling::Types::ChargeResponseChargeModel }, optional: false, nullable: false, api_name: "chargeModel"
      field :billing_timing, -> { Novabilling::Types::ChargeResponseBillingTiming }, optional: false, nullable: false, api_name: "billingTiming"
      field :invoice_display_name, -> { String }, optional: true, nullable: false, api_name: "invoiceDisplayName"
      field :min_amount_cents, -> { Integer }, optional: true, nullable: false, api_name: "minAmountCents"
      field :prorated, -> { Internal::Types::Boolean }, optional: false, nullable: false
      field :properties, -> { Internal::Types::Hash[String, Object] }, optional: true, nullable: false
      field :graduated_ranges, -> { Internal::Types::Array[Novabilling::Types::ChargeGraduatedRangeResponse] }, optional: false, nullable: false, api_name: "graduatedRanges"
      field :filters, -> { Internal::Types::Array[Novabilling::Types::ChargeFilterResponse] }, optional: false, nullable: false
      field :created_at, -> { String }, optional: false, nullable: false, api_name: "createdAt"
      field :updated_at, -> { String }, optional: false, nullable: false, api_name: "updatedAt"
    end
  end
end
